mysql-test/suite/innodb/t/percona_bulk_load_insert.test

```diff
@@ -0,0 +1,65 @@
+--source include/have_debug.inc
+--source include/have_debug_sync.inc
+
+SET GLOBAL innodb_fill_factor = 10;
+SET GLOBAL innodb_semaphore_wait_timeout_debug = 120;
+SET GLOBAL DEBUG = '+d, innodb_bulk_load_compress_sleep';
+--source include/count_sessions.inc
+
+connect (con1,localhost,root,,);
+connect (con2,localhost,root,,);
+
+CREATE TABLE t1(
+        class   INT,
+        id      INT,
+        title   VARCHAR(100),
+	title2  VARCHAR(100)
+) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4;
+
+# Create Insert Procedure
+DELIMITER |;
+CREATE PROCEDURE populate_t1()
+BEGIN
+        DECLARE i int DEFAULT 1;
+
+        START TRANSACTION;
+        WHILE (i <= 30000) DO
+                INSERT INTO t1 VALUES (i, i, uuid(), uuid());
+                SET i = i + 1;
+        END WHILE;
+        COMMIT;
+END|
+DELIMITER ;|
+
+--disable_query_log
+CALL populate_t1();
+--enable_query_log
+
+--connection con2
+SET DEBUG_SYNC = 'bulk_load_insert SIGNAL alter_started WAIT_FOR resume_alter';
+--send CREATE INDEX idx_title ON t1(title, title2);
+
+--connection con1
+SET DEBUG_SYNC = 'now WAIT_FOR alter_started';
```

Comment:
I have trouble visualizing the thread sync points, especially with non-debug_sync innodb_bulk_load_compress_sleep. Could you please add comments?

Reply:
done added this comment:

# This statement helps to run ALTER for long time without actually
# creating a very big table
SET GLOBAL DEBUG = '+d, innodb_bulk_load_compress_sleep';

Reply:
But why do we need to do it non-debug_sync? If I understand correctly, it hits after "resume_alter" sync point? Can't we just hold ALTER TABLE thread at where innodb_bulk_load_compress_sleep is placed with another debug_sync point?

Reply:
no, it is not about synchronization. We want ALTER to run for long time. Thats the only objective with the sleep.

Lets say if we have 10 million mtrs (PageBulks) during ALTER, and since all mtrs/PageBulks hold index->lock, the total duration of index->lock acquired by ALTER crosses 600 seconds.

It is possible to run without this sleep, then as seen in testcase, we will have need around 30GB datafile and also preferably slow CPUs.

To avoid all these huge requirements and slow CPUs, a sleep is used to simulate a "long running ALTER".

Reply:
To add further, mtr with sleep takes 1sec, So we only need 600 mtrs/Pagebulks.

Reply:
But doesn't debug_sync substitute the long time need as well?
Consider: without the fix, ALTER holds the X latch, and would be stopped with debug_sync
INSERT tries to S-lock, but will be prevented for arbitrary long time due to debug_sync-paused ALTER
With the fix, the ALTER can proceed the moment INSERT finishes, no need to cross 600 secs neither

Reply:
I understand, it is like stopping the thread after index->lock acquisition in ALTER and then INSERT would get blocked without the fix. With the fix, INSERT wouldn't get blocked.

I dont see it as close to real scenario. With debug sync point after index->lock acquisition, we could reproduce even with one mtr.

But the peculiar scenario  here is mtrs are short and they are committed (release the latch), yet the index->lock is held more than 600 sec.  The chain of mtrs is special.

Reply:
OK, now I understand